### DISCUSSION TITLE
QST:distance function providing wrong output

### DISCUSSION THREAD

**Discussion by emilianosantin** [geopandas/geopandas#2705] [Q&A]
- [ ] I have searched the [geopandas] tag on [StackOverflow](https://stackoverflow.com/questions/tagged/geopandas) and [GIS StackExchange](https://gis.stackexchange.com/questions/tagged/geopandas) for similar questions.

- [ ] I have asked my usage related question on [StackOverflow](https://stackoverflow.com) or [GIS StackExhange](https://gis.stackexchange.com).

---

#### 
Hi 
I am trying to calculate the minimum distance between points in a shapefile and other geometries  lines. In the results a get several 0 values and other quite different from manual measurements using the measuring tool. I know that layers don't perfectly align but I cannot get my head around the 0 values. I double checked as you can see both files were converted to the same crs.

Thanks

Emiliano 

```python
import os
import geopandas as gpd 

import pandas as pd
from shapely import speedups

speedups.disable()

points_50 = pd.read_csv('data.csv')

points_50_gdf = gpd.GeoDataFrame(
    points_50, crs='EPSG:3857', geometry=gpd.points_from_xy(points_50.longitude, points_50.latitude))

points_50_gdf.crs
<Projected CRS: EPSG:3857>
Name: WGS 84 / Pseudo-Mercator
Axis Info [cartesian]:
- X[east]: Easting (metre)
- Y[north]: Northing (metre)
Area of Use:
- name: World between 85.06°S and 85.06°N.
- bounds: (-180.0, -85.06, 180.0, 85.06)
Coordinate Operation:
- name: Popular Visualisation Pseudo-Mercator
- method: Popular Visualisation Pseudo Mercator
Datum: World Geodetic System 1984
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

wtr_ar_list = []

rootdir = os.getcwd()

for subdir, dirs, files in os.walk(rootdir):
    for file in files:
        filepath = subdir + os.sep + file
        if filepath.endswith("WTR_AR.shp"):
           wtr_ar_list.append(filepath)

gdf = gpd.GeoDataFrame(pd.concat([gpd.read_file(i) for i in wtr_ln_list], 
                        ignore_index=True), crs='EPSG:3857')
<ipython-input-41-fbc1c905590c>:1: FutureWarning: CRS mismatch between CRS of the passed geometries and 'crs'. Use 'GeoDataFrame.set_crs(crs, allow_override=True)' to overwrite CRS or 'GeoDataFrame.to_crs(crs)' to reproject geometries. CRS mismatch will raise an error in the future versions of GeoPandas.



gdf.to_crs(epsg=3857, inplace=True)
<Projected CRS: EPSG:3857>
Name: WGS 84 / Pseudo-Mercator
Axis Info [cartesian]:
- X[east]: Easting (metre)
- Y[north]: Northing (metre)
Area of Use:
- name: World between 85.06°S and 85.06°N.
- bounds: (-180.0, -85.06, 180.0, 85.06)
Coordinate Operation:
- name: Popular Visualisation Pseudo-Mercator
- method: Popular Visualisation Pseudo Mercator
Datum: World Geodetic System 1984
- Ellipsoid: WGS 84
- Prime Meridian: Greenwich

def min_distance(point, gdf):
    return gdf.distance(point).min()

points_50_gdf['dtc_ln_wtr'] = points_50_gdf.geometry.apply(min_distance, args=(gdf,))

```

**Comment by martinfleis** [comment]
Are you sure you are setting CRS properly? Notice the warning: `CRS mismatch between CRS of the passed geometries and 'crs'.`. Without seeing your data it is a bit hard to help, can you try to provide a minimal example following [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports)?

0 might mean that the point intersects the line.

**Comment by martinfleis** [comment]
> It is me again. I am trying to find the minimum distance between points and the coastline. I am using a shapefile from NOAA nos80k.shp http://pubs.usgs.gov/of/2012/1005/data/basemaps/nos80k/nos80k.zip to measure between points and the shape file. Most of the points are inside the shapefile, some even in the mid plains far enough from the coast and I get 0 values. I tried to different ways to iterate through the rows (answers in stackoverflow). Again I am not sure why I get is zero values even if they far away from the coastline, both files have been reprojected to the same crs. Some points return values that are clearly wrong from their distance to the coast. Thanks

If you are trying to get a distance to the boundary of a Polygon not the Polygon itself, you have to get the boundary first.

```py
boundary = gdf.boundary

points_50_gdf['dtc_ln_wtr'] = points_50_gdf.geometry.apply(min_distance, args=(boundary,))
```

**Comment by emilianosantin** [comment]
Thank you! I will check that and get back to you.

**Comment by emilianosantin** [comment]
Another quick question. The script takes quite long. Is there any geopandas function to limit the distance() iterations to closest neighbors? 

Thank you again for taking the time to look at my question

Emiliano

**Comment by martinfleis** [comment]
You should use spatial index to find the nearest neighbour and its distance. @brendan-ward wrote a great tutorial here https://medium.com/@brendan_ward/how-to-leverage-geopandas-for-faster-snapping-of-points-to-lines-6113c94e59aa which will surely help you to write the solution you need.
